src/servers/http_server_v2.cc

```diff
@@ -413,13 +413,55 @@ HTTPAPIServerV2::ResponseRelease(
   return nullptr;  // Success
 }
 
+template <typename T>
+void
+ReadDataFromJsonHelper(
+    std::vector<T>* data_vec, const char* dtype,
+    const rapidjson::Value& payload_data, int* counter)
+{
+  for (size_t i = 0; i < payload_data.Size(); i++) {
+    // If last dimension
+    if (!payload_data[i].IsArray()) {
+      if (strcmp(dtype, "BOOL")) {
```

Comment:
This comparison tree and others like it are very inefficient. You should convert the dtype string once into DataType and then pass that around and use as necessary with a switch. In common.cc you can rename GetDataTypeProtocolString to DataTypeToProtocolString and add ProtocolStringToDataType. In ProtocolStringToDataType instead comparing the entire string over and over break it into a search tree where you look at just the first letter, etc. so that you only have to do 1 full string compare at the end to verify.

Reply:
Done.